### PR TITLE
Fix streaming of big files from Azure Blob Storage

### PR DESCRIPTION
**Pull request by amireluk**
Fix issue #307:'Unable to read beyond the end of the stream.' 
The fix allows streaming big files from Azure Blob Storage

### Description
Due to the way Azure blob are streamed, when streaming a big file the stream length will be correct and the reader will know how many bytes to expect but when making a 'read' request only part of the requested bytes are actually being read (resulting in "Unable to read beyond the end of the stream" error).
This can be solved with consecutive 'read' actions until the entire expected amount it read.

- [x] I have **not** included unit tests validating this fix.
(no UT added, the fix is for streaming big files from Azure, not UT material. But it was tested locally and in a running system in Azure)
- [x] I have updated markdown documentation where required.
- [x] I understand that successful approval of my pull request requires reproducible tests as per [Contribution Guideline](https://github.com/elastacloud/parquet-dotnet/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: src/Parquet/File/DataStreamFactory.cs [hunk 1]
```diff
@@ -62,12 +62,20 @@ static class DataStreamFactory
...
          byte[] data = BytesPool.Rent(compressedLength);
-         int read = nakedStream.Read(data, 0, compressedLength);
 
-         if(read != compressedLength)
+         // Some storage solutions (like Azure blobs) might require more than one 'Read' action to read the requested length.
+         while (totalBytesRead < compressedLength)
          {
-            throw new ParquetException($"expected {compressedLength} bytes in source stream but could read only {read}");
+            currentBytesRead = nakedStream.Read(data, totalBytesRead, compressedLength - totalBytesRead);
+            totalBytesRead += currentBytesRead;
```
Comment (CurtHagenlocher) [inline src/Parquet/File/DataStreamFactory.cs]: This should have "if (currentBytesRead == 0)" { break; }"

Comment (amireluk) [reply src/Parquet/File/DataStreamFactory.cs]: added "&& currentBytesRead != 0" to the while condition.
and change currentBytesRead init to int.min for first iteration.

Comment (amireluk) [reply src/Parquet/File/DataStreamFactory.cs]: @CurtHagenlocher 
Can you approve this PR or should I add someone else for that?

Comment (CurtHagenlocher) [reply src/Parquet/File/DataStreamFactory.cs]: Sorry...

Comment (amireluk) [reply src/Parquet/File/DataStreamFactory.cs]: That was an honest question, didn't mean to be rude :)
I'm still not sure what's the process

**Comment by CurtHagenlocher** [comment]
I meant that „sorry I can’t commit it because it looks right to me.“

**Comment by amireluk** [comment]
@aloneguid

**Comment by aloneguid** [comment]
TBH not very keen to accept it as it seems like a workaround for one SDK that reports stream length wrong. It could be better to include this in some sort of FAQ here on the website.

Also when you have massive parquet files which workaround will actually crash with OOM.

**Comment by CurtHagenlocher** [comment]
It has nothing to do with the reported length; it's that the contract on Stream.Read is that it can return less than the number of bytes requested -- and that this doesn't mean that the end of the stream has been reached.

**Comment by aloneguid** [comment]
got you, apologies

**Pull request merged**